File: Runtime/DeterministicTransitionMap.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.Core

Class DeterministicTransitionMap
    Inherits STSimMapBase2(Of DeterministicTransition)

    Public Sub New(
        ByVal scenario As Scenario,
        ByVal transitions As DeterministicTransitionCollection)

        MyBase.New(scenario)

        For Each t As DeterministicTransition In transitions
            AddTransition(t)
        Next

    End Sub

    Public Function GetDeterministicTransition(
        ByVal stratumId As Nullable(Of Integer),
        ByVal stateClassId As Integer,
        ByVal iteration As Integer,
        ByVal timestep As Integer) As DeterministicTransition

        Return Me.GetItem(stratumId, stateClassId, iteration, timestep)

    End Function

    Private Sub AddTransition(ByVal t As DeterministicTransition)

        Try

            Me.AddItem(
                t.StratumIdSource,
                t.StateClassIdSource,
                t.Iteration,
                t.Timestep,
                t)

        Catch ex As STSimMapDuplicateItemException

            Dim template As String =
                "A duplicate deterministic transition was detected: More information:" & vbCrLf &
                "Source {0}={1}, Source State Class={2}, Iteration={3}, Timestep={4}."

            ExceptionUtils.ThrowArgumentException(
                template,
                Me.PrimaryStratumLabel,
                Me.GetStratumName(t.StratumIdSource),
                Me.GetStateClassName(t.StateClassIdSource),
                STSimMapBase.FormatValue(t.Iteration),
                STSimMapBase.FormatValue(t.Timestep))

        End Try

        Debug.Assert(Me.HasItems())

    End Sub

End Class

File: src/Runtime/DeterministicTransitionMap.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2023 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using System;
using SyncroSim.Core;
using System.Diagnostics;

namespace SyncroSim.STSim
{
    internal class DeterministicTransitionMap : STSimMapBase2<DeterministicTransition>
    {
        public DeterministicTransitionMap(Scenario scenario, DeterministicTransitionCollection transitions) : base(scenario)
        {

            foreach (DeterministicTransition t in transitions)
            {
                AddTransition(t);
            }
        }

        public DeterministicTransition GetDeterministicTransition(int? stratumId, int stateClassId, int iteration, int timestep)
        {
            return this.GetItem(stratumId, stateClassId, iteration, timestep);
        }

        private void AddTransition(DeterministicTransition t)
        {
            try
            {
                this.AddItem(t.StratumIdSource, t.StateClassIdSource, t.Iteration, t.Timestep, t);
            }
            catch (STSimMapDuplicateItemException)
            {
                string template = "A duplicate deterministic transition was detected: More information:" + Environment.NewLine + "Source {0}={1}, Source State Class={2}, Iteration={3}, Timestep={4}.";
                ExceptionUtils.ThrowArgumentException(template, this.PrimaryStratumLabel, this.GetStratumName(t.StratumIdSource), this.GetStateClassName(t.StateClassIdSource), STSimMapBase.FormatValue(t.Iteration), STSimMapBase.FormatValue(t.Timestep));
            }

            Debug.Assert(this.HasItems);
        }
    }
}
